platform/darwin/src/NSPredicate+MGLAdditions.mm

```diff
@@ -0,0 +1,63 @@
+#import "NSPredicate+MGLAdditions.h"
+
+class FilterEvaluator {
+public:
+    
+    mbgl::Value getValue(id obj) {
+        if ([obj isKindOfClass:NSString.class]) {
+            NSString *string = (NSString *)obj;
+            return std::string(string.UTF8String);
+        } else if ([obj isKindOfClass:NSString.class]) {
+            NSNumber *number = (NSNumber *)obj;
+            if((strcmp([number objCType], @encode(int))) == 0) {
+                return number.intValue;
+            } else if ((strcmp([number objCType], @encode(float))) == 0) {
+                return number.floatValue;
+            } else {
+                return number.boolValue;
+            }
+        }
+        return "";
+    }
+    
+    mbgl::style::Filter operator()(NSComparisonPredicate *predicate) {
+        const NSPredicateOperatorType operatorType = predicate.predicateOperatorType;
+        switch (operatorType) {
+            case NSEqualToPredicateOperatorType: {
+                auto filter = mbgl::style::EqualsFilter();
+                filter.key = predicate.leftExpression.keyPath.UTF8String;
+                filter.value = getValue(predicate.rightExpression.constantValue);
+                return filter;
+            }
+            case NSNotEqualToPredicateOperatorType: {
+                auto filter = mbgl::style::NotEqualsFilter();
+                filter.key = predicate.leftExpression.keyPath.UTF8String;
+                filter.value = getValue(predicate.rightExpression.constantValue);
+                return filter;
+            }
+            default:
+            {
+                [NSException raise:@"Predicate operator type not handled" format:@""];
+                break;
+            }
+        }
+        return mbgl::style::NoneFilter();
+    }
+};
+
+@implementation NSPredicate (MGLAdditions)
+
+- (mbgl::style::Filter)mgl_filter
+{
+    if ([self isKindOfClass:NSCompoundPredicate.class]) {
+        [NSException raise:@"Nested predicates (NSCompoundPredicate) are not yet supported" format:@""];
+    }
+    
+    NSComparisonPredicate *predicate = (NSComparisonPredicate *)self;
+    FilterEvaluator evaluator;
+    auto filter = evaluator(predicate);
```

Comment:
A C++ evaluator may not be a great tool for converting from NSPredicate to `mbgl::style::Filter` after all. We generally use this pattern with `mapbox::util::variant`, not an Objective-C umbrella class. Instead, let’s make this method do nothing but assert, then implement `-mgl_filter` overrides in categories on NSComparisonPredicate and NSCompoundPredicate.

The logic for converting a key or value to an `mbgl::Value` should live in a category on NSExpression.

A C++ evaluator remains a good idea for efficiently and compactly converting an `mbgl::style::Filter` into an NSPredicate. We could have the evaluator call into category initializers on NSComparisonPredicate, NSCompoundPredicate, and NSExpression, if that would make the code easier to understand.